### ISSUE TITLE
AWSAppSyncSubscriptionWatcher .cancel() is not work

### ISSUE THREAD

**Issue by Umetomo** [awslabs/aws-mobile-appsync-sdk-ios#82]
**Describe the bug**
Even if you call cancel () on AWSAppSyncSubscriptionWatcher, Memory is not released.

**To Reproduce**
I am creating a chat application
I was using the same UIViewController in several rooms
When entering the room Initialize with

messageWatcher?.cancel()
messageWatcher = nil

And instantiating subscription with

messageWatcher = try appSyncClient.subscribe

As I enter the room many times, the number of instances continues to increase,
Memory usage will continue to increase

Looking at Xcode's Debug Memory Graph, it is as follows

<img width="449" alt="____________________________2018-10-12_19 13 52" src="https://user-images.githubusercontent.com/15145830/46864964-cefd0d80-ce56-11e8-9270-a5576dd04e5e.png">

Please tell me if there is a solution.

**Environment(please complete the following information):**
 - AppSync SDK Version: 2.6.21
 - Dependency Manager: None
 - Swift Version : 4.1

**Device Information (please complete the following information):**
 - Device: iPhoneX
 - iOS Version: iOS 11.4.1

**Additional context**
 - Using Xcode10

**Comment by Umetomo** [comment]
@muellerfr 
Thank you for your reply.
As .cancel () often crashes
Please tell me the solution

**Comment by Umetomo** [comment]
I have discovered some of the factors that cause clashes.

My project uses a pod with Fabric, but the number of times it has recently climbed increased when I recently upgraded it.

However, I suspect that the Fabric is related to the StopSubscription class, and when I removed the Fabric from the project, it started to work more stable.

As a matter of fact, the AWSCore library has been using Fabric internally
This may be the cause.

In particular, if you have the latest version of Fabric (1.8.0) and Crashlytics (3.11.0)
iPhoneX iOS11.4.1 clashed at a high frequency.

My project uses Cognito authentication.

**Comment by rohandubal** [comment]
Hello @Umetomo 

On closer inspection, I found that there is a case in the SDK where you cancel a subscription before it is activated, there might be some memory leak. I am working on changes to avoid that scenario by handling the cancel call better. 

As a best practice in the meanwhile, you can set the subscriptionWatcher to nil instead of the .cancel() method since the deinit internally calls the cancel method. You should see better performance in that case. 

Regarding Fabric, the AWSCore does have Fabric in it, but they are just the headers. There is no implementation in it. The headers are only used for method references. The under-lying implementation (one from the app's dependency) is the primary source from which these method implementations would be used. What was the version of Fabric you were using?

AppSync SDK and subscriptions do not use any critical part of Core and the issue might be some injection from Fabric/Crashlytics. 

Hope this helps.

Thanks,
Rohan

**Comment by rohandubal** [comment]
Hello @Umetomo 

We have released SDK version `2.6.23` which addresses some memory leaks and retain cycles. Please try it out and let us know if you still face the issues.

Thanks,
Rohan

**Comment by Umetomo** [comment]
@rohandubal 

Thank you for resolving the memory leak.
I am surprised that you responded so quickly!

I tried the new SDK. Wonderfully it did not leak memory!
I feel that the operation is more stable than before.
However, as the problem of crashing still occurs
I will continue to support crash log sharing.

My project uses multi subscription
At the timing of UIApplicationWillResignActive, we set nil to all subscriptions. It seems that there are many things that will crash at this timing.


Fabric seems to have crashed independently without regard to AppSync
Thank you for the analysis.
